### DISCUSSION TITLE
Help Needed with Automating User Invitations via Vaultwarden API

### DISCUSSION THREAD

**Discussion by Gecko-with-a-hat** [dani-garcia/vaultwarden#5054] [Q&A]
Hello everyone,

I'm trying to automate user invitations using the Vaultwarden API. Despite my efforts to research, I keep encountering authorization errors. I can see from the server logs that the API requests are reaching the server. However, I'm unsure how to authorize correctly since I'm not very experienced with programming or API requests.

Here is my code:
```
import requests
import json

admin_token = "MyToken"
email_invite = "user@example.com"

payload = {
    "email": email_invite
}

headers = {
    'Content-Type': 'application/json',
    'Authorization': f'Bearer {admin_token}'
}

response = requests.post("https://vaultwarden.mydomain.com/admin/invite", 
                         headers=headers, 
                         json=payload)

print(response.json())
```

This is the response I receive:
```
{'error': '', 'errorModel': {'message': 'Authorization failed.', 'object': 'error'}, 'error_description': '', 'exceptionMessage': None, 'exceptionStackTrace': None, 'innerExceptionMessage': None, 'message': 'Authorization failed.', 'object': 'error', 'validationErrors': {'': ['Authorization failed.']}}
```

I know the request is reaching the server, as indicated by the following log entries:
```
[request][INFO] POST /admin/invite
[vaultwarden::api::admin][ERROR] Authorization failed.
[response][INFO] (invite_user) POST /admin/invite => 401 Unauthorized
```

The admin token I’m using is the same one I log in with on the admin webpage for my server. What am I doing wrong?

Kind regards,
Gecko

**Comment by Gecko-with-a-hat** [comment]
With help of [this](https://www.reddit.com/r/vaultwarden/comments/1fx9vy3/comment/lqnh2mk/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) reddit comment I have figure out that I need a cookie and JWT to authorize. I have simplified and edited my python script to:

```
import requests
import json

response = requests.post("https://vaultwarden.mydomain.com/admin/invite", 
                         headers={'Content-Type': 'application/json'}, 
                         json={"email": "user@example.com"},
                         cookies={'Admin-Token': "MyToken"})

print(response.status_code)
print(response.json())
```

I am still truly unsure what I need to authenticate. No matter what I can come up with to use as the admin token I keep getting the 401 response and the same error in the response. Is the admin token the same as the password I use to log into the admin WebUI? Do I need to use the cookie that I can find after logging into the WebUI when I press F12 ?
How do I convert these to a JWT? I have tried using ChatGPT to figure it out and I still am completely unsure on what to do. AFAICT there is no documentation on how to use the API? If there is any, could somebody please refer me to it.

Thanks in advance for any help given!

**Reply by caxap647** [reply]
Break the request into 2 stages:
1. POST request to /admin with token param for authorization. On successful authorization you will receive cookies
2. POST request to /admin/invite with cookie to invite a user

In my case, this option worked without errors.

**Reply by Gecko-with-a-hat** [reply]
Thanks for your reply. I have decided to use Bitwarden Directory Connector for my automation.
But I am curious, how would I go about doing that first part? How would that request look like?
